### PR TITLE
Llm string prep

### PR DESCRIPTION
**Pull request by reeceshuttle**
This works by passing in the standard pddl problem and will first strip leading and trailing while space, then adjust lines with only ')' characters on them, and finally remove empty lines.

### CODE REVIEW

File: llm4pddl/utils.py [hunk 1]
```diff
@@ -176,6 +176,33 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
     return Task(domain_file, problem_file)
 
 
+def prep_problem(problem: str) -> str:
```
Comment (tomsilver) [inline llm4pddl/utils.py:179]: let's call this something more descriptive -- maybe `minify_pddl_problem`?

File: llm4pddl/utils.py [hunk 2]
```diff
@@ -176,6 +176,33 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
...
+    """Maps a string of a problem file to a new string.
+
+    This new string has had all space removed between grounded atoms.
```
Comment (tomsilver) [inline llm4pddl/utils.py:182]: add the other things that you described in the PR description to this docstring too

File: llm4pddl/utils.py [hunk 3]
```diff
@@ -176,6 +176,33 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
...
+    new.reverse()
+    for i in range(len(new)):
+        piece = new[i]
+        while len(piece) > 0:
+            if piece[0] == ')':
+                new[i] = piece[1:]
+                piece = piece[1:]
+                new[i + 1] = new[i + 1] + ')'
+            else:
+                print('entered!!!!!!!!!!!!')
```
Comment (tomsilver) [inline llm4pddl/utils.py:197]: remove the print

File: llm4pddl/utils.py [hunk 4]
```diff
@@ -176,6 +176,33 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
...
 
+def prep_problem(problem: str) -> str:
+    """Maps a string of a problem file to a new string.
+
+    This new string has had all space removed between grounded atoms.
+    This is for use before querying llm in order to reduce tokens.
+    """
+    # Gets rid of leading and trailing spaces in every line:
+    new = [piece.strip() for piece in problem.split('\n')]
+    # Moves ')' characters to previous line.
```
Comment (tomsilver) [inline llm4pddl/utils.py:187]: this is a good idea, but i wonder if we can write the code in a simpler way. would the following accomplish the same thing?
`new = new.replace("\n)", ")")`
or something like that?

File: llm4pddl/utils.py [hunk 5]
```diff
@@ -176,6 +176,33 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
...
+        while len(piece) > 0:
+            if piece[0] == ')':
+                new[i] = piece[1:]
+                piece = piece[1:]
+                new[i + 1] = new[i + 1] + ')'
+            else:
+                print('entered!!!!!!!!!!!!')
+                break
+    new.reverse()
+    # Removes lines with empty space
```
Comment (tomsilver) [inline llm4pddl/utils.py:200]: you might want to do this before the `replace` thing that I suggested, to handle cases where there are multiple new lines before a right parentheses

File: llm4pddl/utils.py [hunk 6]
```diff
@@ -176,6 +176,33 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
...
 
 
+def prep_problem(problem: str) -> str:
+    """Maps a string of a problem file to a new string.
+
+    This new string has had all space removed between grounded atoms.
+    This is for use before querying llm in order to reduce tokens.
+    """
+    # Gets rid of leading and trailing spaces in every line:
+    new = [piece.strip() for piece in problem.split('\n')]
```
Comment (tomsilver) [inline llm4pddl/utils.py:186]: let's rename this variable something more descriptive, for example, to make clear that it's a list of strings

File: tests/test_utils.py [hunk 7]
```diff
@@ -202,6 +202,46 @@ def test_get_custom_task():
     assert "Task not found" in str(f)
 
 
+def test_prep_problem():
+    """Tests prep_problem().
+
+    Note: I separately ran prep_problem() and then planned with the
+    new problem. Pyperplan still planned correctly with it.
+    """
+    task01_path = utils.get_custom_task('dressed', 1).problem_file
```
Comment (tomsilver) [inline tests/test_utils.py:211]: nice test! for a function like this with a bunch of tricky edge cases, I think we should write a few more tests. 

can you write some tests that just involve a manually constructed string input to the new function? (so, it doesn't have to come from an actual PDDL file, and it doesn't even necessarily need to be an entire PDDL problem). try to think of tricky edge cases.

**Review by tomsilver — COMMENTED** [review]
great work! this is going to be super nice to have, and hopefully will help us with the token bottleneck we're experiencing right now.

File: llm4pddl/utils.py [hunk 8]
```diff
@@ -176,6 +176,34 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
     return Task(domain_file, problem_file)
 
 
+def minify_pddl_problem(problem: str) -> str:
+    """Maps a string of a problem file to a new string.
+
+    This is for use before querying llm in order to reduce tokens.
+    This works by:
+    1. Gettign rid of space between right parentheses.
```
Comment (tomsilver) [inline llm4pddl/utils.py:184]: Typo: Gettign

File: llm4pddl/utils.py [hunk 9]
```diff
@@ -176,6 +176,34 @@ def get_custom_task(benchmark_name: str, task_num: int) -> Task:
...
+    """
+    # Getting rid of space between right parentheses:
+    prob_wo_space = ')'.join(
+        [piece.strip(' ') for piece in problem.split(')')])
+    # Getting rid of space between left parentheses:
+    prob_wo_space = '('.join([
+        piece.strip(' ') if piece.strip(' ') == '' else piece
+        for piece in prob_wo_space.split('(')
+    ])
+    # Getting rid of leading and trailing whitespace and extra lines after ')':
```
Comment (tomsilver) [inline llm4pddl/utils.py:197]: Is this comment right? It doesn't look like a right parens is used here

File: tests/test_utils.py [hunk 10]
```diff
@@ -202,6 +202,103 @@ def test_get_custom_task():
...
+  (in-closet nice-pants1)
+(in-closet collared-shirt1)
+(in-closet suit-jacket1))
+
+(:goal (and (attending-casual-event person3)
+(attending-formal-event person1)
+(attending-casual-event person4)  
+(attending-formal-event person2) )  )   )"""
+    assert utils.minify_pddl_problem(initial) == answer01
+    # Testing many blank lines:
```
Comment (tomsilver) [inline tests/test_utils.py:282]: Great tests!

**Review by tomsilver — APPROVED** [review]
Looks excellent! Great work

**Comment by tomsilver** [comment]
@reeceshuttle looks like there are some merge conflicts -- let me know if you need help resolving them

**Comment by reeceshuttle** [comment]
I just did `git merge main` and rectified the merge conflicts. pretty sure that worked.

**Comment by reeceshuttle** [comment]
Whenever merging, should the person click `Squash and merge`?

**Comment by tomsilver** [comment]
> Whenever merging, should the person click `Squash and merge`?

Yeah after the PR is approved you can do that!
